File: jtms/java/JQRule.java
// N-Queens rules for JTRE.
// Translated from jqrule.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.
package jtms;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * JQRule registers the queen capture detection rules.
 */
public class JQRule {

    /** Register queens rules. Corresponds to jqrule.lisp. */
    public static void registerQueensRules(Jtre jtre) {
        // Declare Queens-capture as contradiction
        JData.contradiction(Arrays.asList("Queens-capture"), jtre);

        // Register capture detection rule
        Dbclass queenDbclass = JData.getDbclass(Arrays.asList("Queen"), jtre);
        JRules.insertRule(queenDbclass,
            // Matcher
            p -> {
                if (!(p instanceof List)) return new Rule.MatchResult(false, null, false);
                List<?> ps = (List<?>) p;
                if (ps.size() != 3 || !"Queen".equals(ps.get(0)))
                    return new Rule.MatchResult(false, null, false);
                List<Object> bindings = new ArrayList<>();
                bindings.add(ps.get(1)); // column
                bindings.add(ps.get(2)); // row
                return new Rule.MatchResult(true, bindings, true);
            },
            // Body
            args -> {
                if (args.size() < 3) return;
                JTMS.TmsNode q1Node = (JTMS.TmsNode) args.get(0);
                int col1 = ((Number) args.get(1)).intValue();
                int row1 = ((Number) args.get(2)).intValue();

                for (Datum datum : queenDbclass.getFacts()) {
                    if (!(datum.getLispForm() instanceof List)) continue;
                    List<?> qs = (List<?>) datum.getLispForm();
                    if (qs.size() != 3 || !"Queen".equals(qs.get(0))) continue;
                    int col2 = ((Number) qs.get(1)).intValue();
                    int row2 = ((Number) qs.get(2)).intValue();
                    if (col1 == col2) continue;
                    if (JQueens.queensOkay(col1, row1, col2, row2)) continue;
                    JTMS.TmsNode q2Node = datum.getTmsNode();
                    JTMS.TmsNode captureNode = JData.referent(
                        Arrays.asList("Queens-capture"), true, jtre).getTmsNode();
                    jtre.getJtms().justifyNode(
                        Arrays.asList("Death", q1Node, q2Node),
                        captureNode,
                        Arrays.asList(q1Node, q2Node));
                }
            });
    }
}

File: jtms/go/jqrule.go
// N-Queens rules for JTRE.
// Translated from jqrule.lisp in "Building Problem Solvers"
// by Kenneth D. Forbus and Johan de Kleer.

package jtms

// RegisterQueensRules sets up the contradiction and capture rules
// for the N-Queens puzzle.
// Corresponds to the jqrule.lisp file which declares:
//   (contradiction 'Queens-capture *jtre*)
//   (rule ((:IN (Queen ?column1 ?row1) :VAR ?Q1)
//          (:IN (Queen ?column2 ?row2) :VAR ?Q2
//            :TEST (not (or (= ?column1 ?column2)
//                           (queens-okay? ...)))))
//         (rassert! Queens-capture (Death ?Q1 ?Q2)))
func RegisterQueensRules(jtre *Jtre) {
	// Declare Queens-capture as a contradiction
	JContradiction([]interface{}{"Queens-capture"}, jtre)

	// Register the queen-capture detection rule.
	// When two queens are placed and they attack each other,
	// assert the Queens-capture contradiction.
	queenDbclass := JGetDbclass([]interface{}{"Queen"}, jtre)

	InsertJRule(queenDbclass,
		// Matcher: check if this is a (Queen col row) fact
		func(p interface{}) (bool, []interface{}, bool) {
			ps, ok := p.([]interface{})
			if !ok || len(ps) != 3 || ps[0] != "Queen" {
				return false, nil, false
			}
			return true, []interface{}{ps[1], ps[2]}, true // need node
		},
		// Body: for each queen placement, check against all others
		func(args ...interface{}) {
			if len(args) < 3 {
				return
			}
			q1Node := args[0].(*TmsNode)
			col1, _ := toInt(args[1])
			row1, _ := toInt(args[2])

			// Check against all other Queen facts
			for _, datum := range queenDbclass.Facts {
				qs, ok := datum.LispForm.([]interface{})
				if !ok || len(qs) != 3 {
					continue
				}
				col2, _ := toInt(qs[1])
				row2, _ := toInt(qs[2])
				if col1 == col2 {
					continue // same column, skip
				}
				if QueensOkay(col1, row1, col2, row2) {
					continue // non-attacking, skip
				}
				q2Node := datum.TmsNode
				// Assert Queens-capture justified by the two queen nodes
				captureNode := JReferent([]interface{}{"Queens-capture"}, true, jtre).TmsNode
				JustifyNode(
					[]interface{}{"Death", q1Node, q2Node},
					captureNode,
					[]*TmsNode{q1Node, q2Node},
				)
			}
		},
	)
}

// toInt converts an interface{} to int.
func toInt(x interface{}) (int, bool) {
	switch v := x.(type) {
	case int:
		return v, true
	case int64:
		return int(v), true
	case float64:
		return int(v), true
	default:
		return 0, false
	}
}
